### PR TITLE
asyncified filters fixes #220 #213 #159

### LINKED ISSUE #220: Make all communication async.
I just adapted my application to work asynchronously using callbacks. The DAPP is way more responsive now. But it still lags sometimes. It could be due to the sha3 calls that are made for every request to get function signatures.

### PR DESCRIPTION
**Pull request by frozeman**
Made all internal requests async to stop interface lagging.

### CODE REVIEW

File: lib/web3/filter.js [hunk 1]
```diff
@@ -83,51 +84,69 @@ var Filter = function (options, methods, formatter) {
...
+    var intervalId = setInterval(function(){
 
-        messages.forEach(function (message) {
-            message = self.formatter ? self.formatter(message) : message;
-            self.callbacks.forEach(function (callback) {
-                callback(null, message);
-            });
-        });
-    };
+        if(self.filterId || self.filterError)
```
Comment (debris) [inline lib/web3/filter.js]: won't work if the filterId === 0

File: lib/web3/filter.js [hunk 2]
```diff
@@ -83,51 +84,69 @@ var Filter = function (options, methods, formatter) {
...
+            clearInterval(intervalId);
 
-    // call getFilterLogs on start
-    if (!utils.isString(this.options)) {
-        this.get(function (err, messages) {
-            // don't send all the responses to all the watches again... just to this one
-            if (err) {
-                callback(err);
+        if(!self.filterId)
```
Comment (debris) [inline lib/web3/filter.js]: won't work if the filterId === 0

File: lib/web3/filter.js [hunk 3]
```diff
@@ -83,51 +84,69 @@ var Filter = function (options, methods, formatter) {
...
+            return;
+
+        self.callbacks.push(callback);
```
Comment (debris) [inline lib/web3/filter.js]: why callback is pushed after 3 seconds?

File: lib/web3/function.js [hunk 4]
```diff
@@ -46,6 +47,12 @@ SolidityFunction.prototype.extractCallback = function (args) {
     }
 };
 
+SolidityFunction.prototype.extractDefaultBlock = function (args) {
```
Comment (debris) [inline lib/web3/function.js]: this is different functionality, please separate it to another pr

File: lib/web3/requestmanager.js [hunk 5]
```diff
@@ -226,9 +226,16 @@ RequestManager.prototype.poll = function () {
         }
 
         results.map(function (result, index) {
-            result.callback = self.polls[index].callback;
-            return result;
+            // make sure the filter is still installed after arrival of the request
+            if(self.polls[index]) {
```
Comment (debris) [inline lib/web3/requestmanager.js]: this is not a solution. `self.polls` is an Array and you may still call callback for different filter.

File: test/contract.js [hunk 6]
```diff
@@ -218,6 +241,31 @@ describe('web3.eth.contract', function () {
...
+                    data: '0x' + sha3(signature).slice(0, 8) + '0000000000000000000000001234567890123456789012345678901234567890',
+                    to: address,
+                    from: address,
+                    gas: '0xc350'
+                }, '0xb']);
+            });
+
+            var contract = web3.eth.contract(desc).at(address);
+
+            var r = contract.balance.call(address, {from: address, gas: 50000}, 11);
```
Comment (debris) [inline test/contract.js]: Looking at the code above, async `calls` won't work. Please fix this and make a test. Also remember to move it to separate pr :)

File: lib/web3/requestmanager.js [hunk 7]
```diff
@@ -226,9 +226,16 @@ RequestManager.prototype.poll = function () {
...
+                result.callback = self.polls[index].callback;
+                return result;
+            } else
+                return false;
         }).filter(function (result) {
+            if(!result)
+                return false;
```
Comment (debris) [inline lib/web3/requestmanager.js]: also, it's usually better and cleaner to make a separate filter for this kind of statements

File: test/contract.js (hunk 6 above)
Comment (debris) [inline test/contract.js]: update*

ok, i see that it will work, but 

```javascript
contract.balance.call(address, {from: address, gas: 50000}, 11, function (result) { } )
contract.balance(address, {from: address, gas: 50000}, 11, function (result) { } )
```

juz doesn't look nice :/

what do you think about?

```javascript
contract.atBlock(11).balance.call(address, {from: address, gas: 50000}, function (result) { } )
contract.atBlock(11).balance(address, {from: address, gas: 50000}, function (result) { } )
```

although it's still not perfect

File: lib/web3/filter.js [hunk 8]
```diff
@@ -83,51 +84,69 @@ var Filter = function (options, methods, formatter) {
...
     var self = this;
 
-    var onMessage = function (error, messages) {
-        if (error) {
-            return self.callbacks.forEach(function (callback) {
-                callback(error);
-            });
-        }
+    // check inf an interval of 10ms if the filter id has arrived
+    var intervalId = setInterval(function(){
```
Comment (debris) [inline lib/web3/filter.js]: please make sure, that `setInterval` is available in go environment. I expect it to be missing.

File: lib/web3/filter.js [hunk 9]
```diff
@@ -83,51 +84,69 @@ var Filter = function (options, methods, formatter) {
...
 
-        messages.forEach(function (message) {
-            message = self.formatter ? self.formatter(message) : message;
-            self.callbacks.forEach(function (callback) {
-                callback(null, message);
-            });
-        });
-    };
+        if(self.filterId || self.filterError)
+            clearInterval(intervalId);
```
Comment (debris) [inline lib/web3/filter.js]: also `clearInterval` may be missing in go jsenv.

File: test/contract.js (hunk 6 above)
Comment (frozeman) [inline test/contract.js]: I would rather keep the above as its in-line with normal `call(options [, defaultBlock] [, callback])` calls. It does look perfect, but its guessable.

File: lib/web3/filter.js (hunk 8 above)
Comment (frozeman) [inline lib/web3/filter.js]: setInterval as setTimeout was implement in otto, otherwise out filters wouldn't work anyway.

File: lib/web3/requestmanager.js [hunk 10]
```diff
@@ -220,14 +229,21 @@ RequestManager.prototype.poll = function () {
...
-            return result;
+            var key = pollsKeys[index];
+            // make sure the filter is still installed after arrival of the request
+            if(self.polls[key]) {
+                result.callback = self.polls[key].callback;
+                return result;
+            } else
+                return false;
+        }).filter(function (result) {
+            return (!result) ? false : true;
```
Comment (debris) [inline lib/web3/requestmanager.js]: minor, could be also:
```javascript
return !!result;
```

**Comment by debris** [comment]
looks much better now! :+1: :+1:
